bpfd/src/bpf.rs

```diff
@@ -311,7 +317,17 @@ impl BpfManager {
                     map_pin_path.expect("map_pin_path must be set after successfult load"),
                 )
                 .await?;
-                Ok(program)
+
+                // add_multi_attach_program() and add_single_attach_program() take ownership
+                // of program, so this function now has an outdated copy (for example, map_used_by
+                // is not set). Get the latest copy to return.
```

Comment:
This comment/diff is incorrect both add_single/multi_attach_program take mutable references to programs (&mut program)

So therefore after attaching map_pin_path and map_used_by should be set. See the existing logic on line 308 where we get the map pin path (which was set in the attach functions)

Reply:
We might need to consider calling save_map BEFORE inserting the program into the dispatcher 

regardless if the `program` variable in `add_program` is not up to date with what actually got inserted into `self.programs` here we're doing something wrong and should try and figure it out.  I really don't want to hack around this as we just spent a bunch of time refactoring the code to avoid getting/resetting self.programs

Reply:
So `map_used_by` was definitely `None`. So in `add_multi_attach_program()`, after the insert, even though we have the program, we get a fresh copy to do the `p.set_attached()` and `p.save(id)`:
```
        self.dispatchers.insert(did, dispatcher);
        let id = program
            .kernel_info()
            .expect("kernel info should be set after load")
            .id;
        self.programs.insert(id, program.to_owned());

        if let Some(p) = self.programs.get_mut(&id) {  <== Why do we do this then?
            p.set_attached();
            p.save(id)
                .map_err(|e| BpfdError::Error(format!("unable to save program state: {e}")))?;
        };
```